### PR TITLE
fixes #184

### LINKED ISSUE #184: Activity source memory leak
When the `HttpClientRequestAdapter` is resolved from a service provider using either transient or scoped lifetime and passing a HttpClient then it leaks memory.  The leak is caused by the creation of an `ActivitySource` every time the `HttpClientRequestAdapter` is constructed but it's never disposed unless the `HttpClientRequestAdapter` created the HttpClient.

https://github.com/microsoft/kiota-http-dotnet/blob/eb76ee893324bd373462e98dfdafeba95762feff/src/HttpClientRequestAdapter.cs#L55
https://github.com/microsoft/kiota-http-dotnet/blob/eb76ee893324bd373462e98dfdafeba95762feff/src/HttpClientRequestAdapter.cs#L604-L608

This problem was introduced in commit https://github.com/microsoft/kiota-http-dotnet/commit/e44fbfbad61e86d449ccc01d1a2207ec432e0e57#diff-d9ae4f1159bbe7195bce06dbc06baeb882e3f3d4fcf280c3d31523bf8c0dada7

In a very simple test constructing 1 million `HttpClientRequestAdapter` objects which can easily happen in the use-case I have results in the following memory usage pattern. Once the test completes we have a large number of objects in gen2 of the GC.

```
    public static void DotPeekActivitySourceLeak()
    {
        HttpClient httpClient = new HttpClient();
        for (var i = 0; i < 1000000; i++)
        {
            using var adapter = new HttpClientRequestAdapter(new AnonymousAuthenticationProvider(), httpClient: httpClient);
        }
    }
```

Screenshot 2023-12-25 at 7 17 22 AM

Updating the `HttpClientRequestAdapter` constructor to register the ActivitySource in the same way as the HttpClient middleware implementations,  since it's using the same ActivitySource name eliminates this leak.

```
activitySource = ActivitySourceRegistry.DefaultInstance.GetOrCreateActivitySource(obsOptions.TracerInstrumentationName);
```

Screenshot 2023-12-25 at 7 25 22 AM

### PR DESCRIPTION
**Pull request by Jacob-Morgan**
fixes #184: Use activity source registry to resolve memory leak

### CODE REVIEW

File: src/HttpClientRequestAdapter.cs [hunk 1]
```diff
@@ -52,7 +52,7 @@ public HttpClientRequestAdapter(IAuthenticationProvider authenticationProvider,
             pNodeFactory = parseNodeFactory ?? ParseNodeFactoryRegistry.DefaultInstance;
             sWriterFactory = serializationWriterFactory ?? SerializationWriterFactoryRegistry.DefaultInstance;
             obsOptions = observabilityOptions ?? new ObservabilityOptions();
-            activitySource = new(obsOptions.TracerInstrumentationName);
+            activitySource = ActivitySourceRegistry.DefaultInstance.GetOrCreateActivitySource(obsOptions.TracerInstrumentationName);
```
Comment (baywet) [inline src/HttpClientRequestAdapter.cs:55]: This only exists for net5 and higher.
I think we'll need to rely on conditional compilation, and properly dispose of the instance ourselves for anything lower
https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activitysource?view=net-8.0

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for making the changes.
In addition to my comments can you:
- bump the patch version in the csproj
- add a new version in the changelog (today's date)

**Comment by Jacob-Morgan** [comment]
@microsoft-github-policy-service agree company="Pharos"

File: src/HttpClientRequestAdapter.cs (hunk 1 above)
Comment (andrueastman) [reply src/HttpClientRequestAdapter.cs:55]: If you mean the `ActivitySource` type here, I think this should still be okay for netstandard as the type is brought in using the explicit package reference in the csproj file. There's also won't be a need to dispose really as the registry assures only one instance for the entire app lifetime.

Comment (baywet) [reply src/HttpClientRequestAdapter.cs:55]: I meant the DefaultInstance singleton on the ActivitySourceRegistry

Comment (andrueastman) [reply src/HttpClientRequestAdapter.cs:55]: I believe the registry works on framework from the tests [here](https://github.com/microsoft/kiota-http-dotnet/blob/40262ef8cc3268af4949db4062d447820eeca404/Microsoft.Kiota.Http.HttpClientLibrary.Tests/Middleware/ActivitySourceRegistryTests.cs#L20). As the [`ActivitySourceRegistry`](https://github.com/microsoft/kiota-http-dotnet/blob/40262ef8cc3268af4949db4062d447820eeca404/src/Middleware/ActivitySourceRegistry.cs#L14) is a custom class in this project, shouldn't the singleton still work without issue? (I suspect there's something I'm still missing here 😅)

Comment (baywet) [reply src/HttpClientRequestAdapter.cs:55]: thanks for clarifying, I got mixed up when searching for the type in the docs... sorry about the confusion.

**Pull request merged**